File: filter.py
"""
Filter module - Filter OSM objects without elevation data
"""
import json


class ElevationFilter:
    """Filter OSM objects based on elevation data presence"""
    
    @staticmethod
    def has_elevation(element):
        """Check if element has elevation tag"""
        if "tags" in element:
            return "ele" in element["tags"]
        return False
    
    @staticmethod
    def is_alpine_hut(element):
        """Check if element is an alpine hut (cabană montană)"""
        if "tags" in element:
            return element["tags"].get("tourism") == "alpine_hut"
        return False
    
    @staticmethod
    def get_coordinates(element):
        """Extract coordinates from element"""
        if element["type"] == "node":
            return element.get("lat"), element.get("lon")
        elif element["type"] == "way" and "center" in element:
            return element["center"].get("lat"), element["center"].get("lon")
        return None, None
    
    def filter_missing_elevation(self, elements):
        """Filter elements that don't have elevation data"""
        missing_ele = []
        
        for element in elements:
            if not self.has_elevation(element):
                lat, lon = self.get_coordinates(element)
                if lat is not None and lon is not None:
                    missing_ele.append(element)
        
        return missing_ele
    
    def prioritize_alpine_huts(self, elements):
        """Separate alpine huts from other elements for priority processing"""
        alpine_huts = []
        others = []
        
        for element in elements:
            if self.is_alpine_hut(element):
                alpine_huts.append(element)
            else:
                others.append(element)
        
        return alpine_huts, others
    
    def filter_data(self, data):
        """Filter all data and prioritize alpine huts"""
        result = {
            "train_stations": [],
            "alpine_huts": [],
            "other_accommodations": []
        }
        
        # Filter train stations
        if "train_stations" in data:
            result["train_stations"] = self.filter_missing_elevation(data["train_stations"])
        
        # Filter accommodations and prioritize alpine huts
        if "accommodations" in data:
            missing_ele = self.filter_missing_elevation(data["accommodations"])
            alpine_huts, others = self.prioritize_alpine_huts(missing_ele)
            result["alpine_huts"] = alpine_huts
            result["other_accommodations"] = others
        
        return result


if __name__ == "__main__":
    # Load raw data
    with open("output/osm_data_raw.json", "r", encoding="utf-8") as f:
        data = json.load(f)
    
    # Filter
    filter_obj = ElevationFilter()
    filtered = filter_obj.filter_data(data)
    
    # Save filtered data
    with open("output/osm_data_filtered.json", "w", encoding="utf-8") as f:
        json.dump(filtered, f, indent=2, ensure_ascii=False)
    
    print(f"Train stations without elevation: {len(filtered['train_stations'])}")
    print(f"Alpine huts without elevation: {len(filtered['alpine_huts'])}")
    print(f"Other accommodations without elevation: {len(filtered['other_accommodations'])}")
    print("Filtered data saved to output/osm_data_filtered.json")

File: src/filter.go
package main

import (
	"fmt"
)

// ElevationFilter filters OSM elements based on elevation and coordinates
type ElevationFilter struct {
	coordExtractor  *CoordinateExtractor
	categorizer     *ElementCategorizer
}

// FilteredData contains categorized OSM elements
type FilteredData struct {
	TrainStations       []OSMElement `json:"train_stations"`
	AlpineHuts          []OSMElement `json:"alpine_huts"`
	OtherAccommodations []OSMElement `json:"other_accommodations"`
}

// NewElevationFilter creates a new elevation filter
func NewElevationFilter() *ElevationFilter {
	return &ElevationFilter{
		coordExtractor:  NewCoordinateExtractor(),
		categorizer:     NewElementCategorizer(),
	}
}

// filterMissingElevation filters elements without elevation data
func (f *ElevationFilter) filterMissingElevation(elements []OSMElement) []OSMElement {
	var result []OSMElement

	for _, element := range elements {
		if !f.categorizer.HasElevation(element) {
			if f.coordExtractor.HasValidCoordinates(element) {
				result = append(result, element)
			}
		}
	}

	return result
}

// prioritizeAlpineHuts separates alpine huts from other elements
func (f *ElevationFilter) prioritizeAlpineHuts(elements []OSMElement) ([]OSMElement, []OSMElement) {
	var alpineHuts []OSMElement
	var others []OSMElement

	for _, element := range elements {
		if f.categorizer.IsAlpineHut(element) {
			alpineHuts = append(alpineHuts, element)
		} else {
			others = append(others, element)
		}
	}

	return alpineHuts, others
}

// FilterData filters OSM data by elevation status and categorizes elements
func (f *ElevationFilter) FilterData(data *OSMData) *FilteredData {
	result := &FilteredData{
		TrainStations:       []OSMElement{},
		AlpineHuts:          []OSMElement{},
		OtherAccommodations: []OSMElement{},
	}

	// Filter train stations
	result.TrainStations = f.filterMissingElevation(data.TrainStations)

	// Filter accommodations and prioritize alpine huts
	missingEle := f.filterMissingElevation(data.Accommodations)
	alpineHuts, others := f.prioritizeAlpineHuts(missingEle)
	result.AlpineHuts = alpineHuts
	result.OtherAccommodations = others

	return result
}

func runFilter() error {
	fmt.Println("\n" + string(repeat('=', 60)))
	fmt.Println("STEP 2: FILTER - Identifying elements without elevation")
	fmt.Println(string(repeat('=', 60)))

	// Load raw data
	var data OSMData
	if err := loadJSON("output/osm_data_raw.json", &data); err != nil {
		return fmt.Errorf("output/osm_data_raw.json not found. Run --extract first: %v", err)
	}

	// Filter
	filter := NewElevationFilter()
	filtered := filter.FilterData(&data)

	// Save filtered data
	if err := saveJSON("output/osm_data_filtered.json", filtered); err != nil {
		return err
	}

	fmt.Printf("\n✓ Train stations without elevation: %d\n", len(filtered.TrainStations))
	fmt.Printf("✓ Alpine huts without elevation: %d (PRIORITY)\n", len(filtered.AlpineHuts))
	fmt.Printf("✓ Other accommodations without elevation: %d\n", len(filtered.OtherAccommodations))
	fmt.Println("✓ Filtered data saved to output/osm_data_filtered.json")

	return nil
}
